[NOX] Add new tracking events for the new Payment settings page

anu-rock opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #54673.

Add tracking events for the new (Reactified) Payment settings page to capture key moments in the user journey related to providers, incentives, live payments, and business location. See the above issue for a comprehensive list of events.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

##### Local

1. Run `localStorage.setItem( 'debug', 'wc-admin:tracks' );` on your site browser console. And look for verbose logs.
2. Enable the new Payment settings page by going to WP Admin > Tools > WCA Test Helper > Features and turning on the `reactify-classic-payments-settings` flag. (requires the WC Beta Tester plugin).
3. Go to WP Admin > WooCommerce > Settings > Payments. Verify that the `wcadmin_settings_payments_pageview event is recorded.
4. Use instructions in https://github.com/woocommerce/woocommerce/pull/53432 to trigger the action incentive for WooPayments. Check the troubleshooting section if you cannot see the incentive. Refreshing the Payment settings page should record `wcadmin_settings_payments_incentive_show` in addition to the pageview event. Accepting the incentive (clicking the "save 10%" button) should record `wcadmin_settings_payments_incentive_accept`. And dismissing (clicking the "no thanks" button) should record `wcadmin_settings_payments_incentive_dismiss`.
5. Click the Enable button for PayPal provider and verify `wcadmin_settings_payments_provider_installed` is recorded after installation of its plugin.
6. Click the Complete Setup button for PayPal and verify `wcadmin_settings_payments_provider_complete_setup_click` is recorded.
7. With WooPayments disabled and deactivated, click its Enable button. You should be redirected to the payment method selection screen. Go back without modifying selection or pressing the Continue button. If you previously did not create a test account, you will see the Complete Setup button. (If you previously created a test account, you will see Enable and Manage buttons. Choose the reset account option from the ellipses menu to delete the test account and show the Complete Setup button). Click it to create a test account (also verify `wcadmin_settings_payments_provider_complete_setup_click` is recorded). Once finished, you will be redirected back to the Payment settings page and shown a "ready to test payments" modal. Click the Continue Store Setup button and verify `wcadmin_settings_payments_continue_store_setup_click` is recorded.
8. Go back to the Payment settings page. You should see the Manage button for WooPayments. Click it and verify that `wcadmin_settings_payments_provider_manage_click` is recorded.
9. Go back to the Payment settings page. From WooPayments' ellipses menu, choose the disable option and verify that `wcadmin_settings_payments_provider_disable_click` and `wcadmin_settings_payments_provider_disable` are recorded.
10. Once WooPayments is disabled, you will see Enable and Manage buttons. This time click WooPayments' Enable button and verify that `wcadmin_settings_payments_provider_enable_click` is recorded.
11. Change the business location from the default United States to something else, say Ukraine. Verify that `wcadmin_settings_payments_business_location_update` is recorded.

<img width="1900" alt="image" src="https://github.com/user-attachments/assets/0cac4bec-a8f2-4cea-a2a0-c748e2a0f9ce" />

##### JurassicNinja

1. Create a new JN site with WooCommerce on this PR's live branch and the WooCommerce Beta Tester plugin (here is a [direct JN create link](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce-beta-tester&woocommerce&branches.woocommerce=add/54673-tracks-events-new-payment-settings))
2. Go to WP Admin > WooCommerce menu item. You will be taken to the WooCommerce profiler. Click "Skip guided setup" and choose `US` as the store's base location country.
3. Go to WooCommerce > Settings > Advanced > WooCommerce.com and turn the `Enable tracking` flag on.
4. Install [Tracks Vigilante](https://github.com/Automattic/tracks-chrome-extension) browser extension and use its log of events.
13. Follow the steps in the Local section above. Instead of browser console, verify that correct events are recorded in Tracks Vigilante.

Note about testing incentive events:

Triggering action incentive is tricky on JN. Follow these steps to trigger switch incentive:

* The Switch incentive has the following eligibility criteria that we need to mimic:
  * store is in a WooPayments-eligibile country
  * store is 90+ days old → use the WC Beta Tester plugin and from Tools > WCA Test Helper > Tools use the “Update WC installation timestamp” option to set a store timestamp more than 90 days in the past
  * store has a PSP enabled (other than WooPayments) → enable an offline payment method
  * store has orders → create a test order (previously import the dummy products)
  * store doesn’t have WooPayments → do not have WooPayments installed.
* Switch the WooCommerce base location to a different country to force a re-fetch.
* To import the dummy products, use the “Import your products” WC onboarding task page.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

anu-rock committed:
Record events for page view, provider installed, incentive shown/accepted/dismissed

anu-rock committed:
Record events for enabling, disabling, setting up a provider

anu-rock commented on plugins/woocommerce/client/admin/client/settings-payments/components/buttons/enable-gateway-button.tsx:
```diff
@@ -124,6 +131,12 @@ export const EnableGatewayButton = ( {
 					// We only need to perform additional logic/redirects if no account connected.
 					if ( ! gatewayState.account_connected ) {
+						// Record the event when user successfully enables a gateway.
+						recordEvent( 'settings_payments_provider_enable', {
```
I am a bit shaky if it's the right place to capture this event.

dmallory42 replied:
We may want to implement this event on the backend. @vladolaru, what do you think?

anu-rock replied:
We'll leave it like this for now as moving this event to backend PHP code will require hunting down the `woocommerce_toggle_gateway_enabled` action, which may not be worth the effort. Slack discussion: p1738751326045739-slack-C0800LPNCET

dmallory42 replied:
On the latest update, this seems to work nicely so I'm happy to leave it like this.

anu-rock committed:
Add changelog

anu-rock committed:
Record event for provider manage button click

anu-rock committed:
Move the pageview event to a more appropriate place

anu-rock committed:
Record event for business location updated event

anu-rock committed:
Record button clicks in WooPayments ready to test modal

anu-rock commented on plugins/woocommerce/client/admin/client/settings-payments/components/buttons/settings-button.tsx:
```diff
@@ -18,25 +24,33 @@ interface SettingsButtonProps {
+	 * Whether the associated provider plugin is being installed.
 	 */
-	installingPlugin: string | null;
+	isInstallingPlugin?: boolean;
```
I took the liberty of renaming this field as it would provide the gateway ID only before the respective plugin's installation (null after that).

dmallory42 replied:
Looks good to me!

anu-rock committed:
Add the first set of tests

anu-rock committed:
Add changelog for the data package

dmallory42 commented:
This is looking great, nice job Anurag! 

One thing I noticed that we may need to check is that I saw the incentive show event fire multiple times when accepting an incentive. In this screenshot, the initial show event, and the the incentive accept event are expected, but then it looks like two more incentive show events get fired which are not expected. Can we check that behaviour? 

<img width="505" alt="Screenshot 2025-02-04 at 14 48 54" src="https://github.com/user-attachments/assets/ff538f63-2a0a-4f10-b2df-f6edc8fc90d2" />

anu-rock committed:
Add tests for the buttons

anu-rock committed:
Prevent multiple recordings of incentive show event

anu-rock commented:
> One thing I noticed that we may need to check is that I saw the incentive show event fire multiple times when accepting an incentive.

@dmallory42 https://github.com/woocommerce/woocommerce/pull/54957/commits/771f3920eb21d8475b1acac7e5f34960bb5e63e8 should fix that. Thanks for testing!

anu-rock commented:
Created a separate PR for missed unit tests: https://github.com/woocommerce/woocommerce/pull/55177

anu-rock committed:
Attach provider enable events to the other Enable button

... the one that installs and activates a provider when its plugin is already installed. The previous Enable button is used to toggle a gateway on or off.

anu-rock committed:
Add test for provider enable click event

dmallory42 reviewed (approved):
This is looking great, thanks for your work on this! I just tested and apart from the known caveats everything works as I expected. I'm happy to 🚢  (via the CFE branch, which will also need to include https://github.com/woocommerce/woocommerce/pull/54834) once you are ready!

anu-rock commented:
> This is looking great, thanks for your work on this! I just tested and apart from the known caveats everything works as I expected. I'm happy to 🚢 (via the CFE branch, which will also need to include #54834) once you are ready!

@dmallory42 Many thanks for reviewing :) The known caveats were fixed in https://github.com/woocommerce/woocommerce/pull/54957/commits/75de083db204d40a965b80e12557c97c0c0d548e 🎉
